### DISCUSSION TITLE
How to generate props tables/docs from .d.ts declarations files?

### DISCUSSION THREAD

**Discussion by slimshreydy** [storybookjs/storybook#20596] [Help]
I have a typescript React library I've been working on that has components stored in `lib/*.tsx` files. When I run `tsc` to compile my typescript, the resulting JS + declaration files are stored in `dist/*.js` and `dist/*.d.ts` respectively.

Let's say one of my ".tsx" files looks like this:

```tsx
export interface ButtonProps {
  /**
   * Desired color for button, designated as a HEX code.
   */
  color: string;
}

export const Button = ({color}: ButtonProps) => {
  return <button style={{"backgroundColor": color}}>Button</button>
}
```

After compiling, my documentation for the `ButtonProps` gets stored in a `dist/Button.d.ts` file while the implementation gets stored in `dist/Button.js`

When attempting to make a storybook with auto generated prop documentation, importing from `lib/*.tsx` in my stories will show the doc strings in the props table as expected:

```tsx
import { Button } from "../lib/Button";
export default {
  title: "Example/Button",
  component: Button
} as ComponentMeta<typeof Button>
const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;
export const RegularButton = Template.bind({});
```

<img width="601" alt="image" src="https://user-images.githubusercontent.com/10700914/212140046-74808efb-1ab3-4751-9675-e512514cbc2b.png">

But if I switch to importing from `../dist/Button`, I lose all documentation:

<img width="597" alt="image" src="https://user-images.githubusercontent.com/10700914/212140327-880092ed-ff82-444e-a466-c0ed5bfb44b5.png">

Any idea why Storybook isn't able to identify that the docs for the props are in the `.d.ts` files? Is there something I have to configure?

My storybook `main.cjs` config looks like this:
```
const path = require("path");

module.exports = {
  stories: ["../stories/**/*.stories.mdx", "../stories/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
    {
      name: "@storybook/addon-postcss",
      options: {
        postcssLoaderOptions: {
          // When using postCSS 8
          implementation: require("postcss"),
        },
      },
    },
  ],
  framework: "@storybook/react",
  core: {
    builder: "@storybook/builder-webpack5",
  },
  webpackFinal: async (config) => {
    // https://github.com/storybookjs/storybook/issues/15335
    config.module.rules.push({
      resolve: { fullySpecified: false },
    });
    return config;
  },
};
```

My project has some restrictions that make it so I can't refer to the `lib/*.tsx` files from a storybook story file when deployed, but I imagine this is also applicable to anyone who attempts to use Storybook with a precompiled external component library too (where they likely only have access to the compiled `.js` and `.d.ts` files).

**Comment by chakAs3** [comment]
Hi @slimshreydy,  why did you switch to import from ./dist, you should not I guess because 

1. we are not supposed to manually import from dist as it is a build output folder 
2. any component parser should load the component from the source file with clear typed script and annotation, as the ./dist most bundle and optimized code so all the comments with be removed.

**Reply by adam-nielsen** [reply]
The problem is that if you are writing a Storybook for an external library, you don't have access to the source code, as only the bundled/optimised code is distributed with the package (and the `.d.ts` file with the types).

Does this mean Storybook cannot be used to document components imported from external libraries?
